### ISSUE TITLE
Can't type square brackets [ ]

### ISSUE THREAD

**Issue by defaltus** [onivim/oni#644]
For some reason I can't type square brackets using Oni 0.2.8 on Windows 7.
The input works on any other window, just inside Oni it does not work.
It only switches from `insert` to `normal` mode when I try to type `[`. 
When typing `]` Oni does nothing.

I use the standard german keyboard layout (in Neovim without Oni everything works fine).

I am currently not using a custom `init.vim`, my `config.js` looks like this
```javascript
module.exports = {
  //add custom config here, such as
  "oni.useDefaultConfig": true,
  "oni.loadInitVim": false,
  "oni.hideMenu": true,
  "editor.fontSize": "14px",
  "editor.fontFamily": "Roboto Mono"
}
```

**Comment by extr0py** [comment]
Thanks for logging the issue, @Balagrio , and all the details. I'm guessing it is related to some difference in the german keyboard layout. I think this is an issue with the mapping / conversion logic in `Keyboard.ts` - perhaps it is mistaking the `[` keycode for Escape.

**Comment by justjoeyuk** [comment]
@extr0py I don't think so. The global code for ESC is 27, which is the keycode mapping. I tried doing a German keyboard layout and testing different representation of the keycode in JS. It all printed out as ESC being pressed. Can't really see a solid solution.

**Comment by ioncodes** [comment]
I'm expiriencing this too. Same problem happens when I paste text while in Insert mode. Maybe it doesn't handle CRLF/LF correctly?

**Comment by extr0py** [comment]
I think this is because we aren't handling the `AltGr` key correctly - depending on the platform it either comes to our input layer as `C+[` or `C+A+[`, so we need to handle this case.

**Comment by riedel** [comment]
I do not quite get the root of the problem. On the german keyboard  `{[]}\` are AltGr combinations of `7890`. Interestingly `{}` work (!), but not `[]\` which all emit the '^\' escape sequence of ESC, so what is the difference here? BTW also `|` and '~' work, but not `²³`. Trying to type e.g. latex without `[]\` on a German keyboard is pretty much a show stopper.

**Comment by bryphe** [comment]
This was fixed in PR #717 (which is in the v0.2.10 release), the `AltGr` keys should behave as expected now for German keyboards. I'll close this, but let me know if you're still seeing any issues.

**Comment by soratobukuroneko** [comment]
Hi, I just tried right now with the v0.2.11 release on a laptop with windows 10 and a spanish keyword. I can't type the opening square bracket ([), but I can type the closing one normaly.

How could I help, get the key code or so?

**Comment by bryphe** [comment]
Hi @soratobukuroneko, 

Thanks for reporting the issue! I was able to repro on my Win 10 machine with spanish keyboard.

You can see the key resolution in the console by running the command `Oni.log.enableVerboseLogging()`:
![image](https://user-images.githubusercontent.com/13532591/31301979-19f490c4-aab3-11e7-9c57-0fcbf83aebdd.png)

It looks like, in this case, the `BracketLeft` is getting resolved as `<Control+Alt+[>` instead of `[` as we'd expect. The `BracketRight` shows a correct resolution of `]`.

We use the `keyboard-layout` module to help resolve the characters - it provides a keymap that tells us how a character resolves.

Somehow, it´s not providing a resolution for `BracketLeft`:
![image](https://user-images.githubusercontent.com/13532591/31302047-8652d294-aab3-11e7-8f90-34acd1402820.png)

I'll log an issue on the `keyboard-layout` repo to track - in the meantime, we could probably hack this by hardcoding the resolution for the case where we're on windows and using the `es-ES` keyboard language. I just hope there aren't many more of these gaps!

**Comment by bryphe** [comment]
Logged an issue at atom/keyboard-layout#36 to track. In the meantime, I'll fix this locally by special casing this.

**Comment by bryphe** [comment]
FYI @soratobukuroneko  - The fix for this is in the latest build - [v0.2.12](https://github.com/bryphe/oni/releases). Feel free to open a new issue if there are any other keyboard-layout problems - we may need to generalize the 'hack' here.

**Comment by soratobukuroneko** [comment]
I tried the v0.2.12 release and the "1.0.256" build from appveyor but the opening bracket keep getting resolved to  `<c-a-[>` instead of  `BracketLeft`

**Comment by MatonAnthony** [comment]
Hello,

On Windows 7 with a Belgian Layout this problem is still present.
Is there a fix, I could use?

**Comment by Julien00859** [comment]
Up. Same problem on the Belgian layout on Windows 10

**Comment by badosu** [comment]
@MatonAnthony @Julien00859 Can you post the output of the Dev Console as outlined by @bryphe above? Are you using 0.3.2?

**Comment by DeltaEvo** [comment]
@badosu it seems that @bryphe only fixed this for es-ES layout https://github.com/onivim/oni/commit/d5d1d2d1057873a09ab312257ab8725e0f877936, i think to make the fix for everyone we can define it when itts not defined, what do you think @bryphe ?

**Comment by DeltaEvo** [comment]
@Julien00859 the key events seems logged with debug log level so try using `Oni.log.enableDebugLogging()`

**Comment by Julien00859** [comment]
Indeed

    [Key event] Code: BracketLeft Key: [ CtrlKey: true ShiftKey: false AltKey: true | Resolution: <a-c-[>
